lib/pos_consensus_controller.go

```diff
@@ -211,3 +211,109 @@ func (cc *ConsensusController) HandleHeader(pp *Peer, msg *MsgDeSoHeader) {
+	// Try to apply the block locally as the new tip of the blockchain
+	successfullyAppliedNewTip, _, missingBlockHashes, err := cc.blockchain.processBlockPoS(
+		block, // Pass in the block itself
+		cc.fastHotStuffEventLoop.GetCurrentView(), // Pass in the current view to ensure we don't process a stale block
+		true, // Make sure we verify signatures in the block
+	)
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error processing block locally: %v", err)
+	}
+
+	// If the incoming block is an orphan, then there's nothing we can do. We return the missing ancestor
+	// block hashes to the caller. The caller can then fetch the missing blocks from the network and retry
+	// if needed.
+	if len(missingBlockHashes) > 0 {
+		return missingBlockHashes, nil
+	}
+
+	// If the block was processed successfully but was not applied as the new tip, then there's nothing left to do.
+	if !successfullyAppliedNewTip {
+		return nil, nil
+	}
+
+	// Fetch the new tip from the blockchain. Note: the new tip may or may not be the input block itself.
+	// It's possible that there was a descendant of the tip block that was previously stored as an orphan
+	// in the Blockchain, and was applied as the new tip.
+	tipBlockHeader := cc.blockchain.BlockTip().Header
+	tipBlockHash, err := tipBlockHeader.Hash()
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error hashing new tip block: %v", err)
+	}
+
+	// Build a UtxoView at the new tip
+	utxoViewAtTipBlock, err := cc.blockchain.getUtxoViewAtBlockHash(*tipBlockHash)
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching UtxoView for tip block: %v", err)
+	}
+
+	// Fetch the validator set at the new tip
+	validatorSetAtTipBlock, err := utxoViewAtTipBlock.GetFullSnapshotValidatorSet()
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching validator set for tip block: %v", err)
+	}
+
+	// Bundle the tip block and validator set together
+	tipBlockWithValidators := consensus.BlockWithValidatorList{
+		Block:         tipBlockHeader,
+		ValidatorList: ValidatorEntriesToConsensusInterface(validatorSetAtTipBlock),
+	}
+
+	// Fetch the remaining safe blocks that are eligible to be extended from by the next incoming tip block
+	safeBlocks, err := cc.blockchain.getSafeBlocks()
+	if err != nil {
+		return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error fetching safe blocks: %v", err)
+	}
+
+	// Fetch the validator sets for the safe blocks
+	safeBlocksWithValidators := make([]consensus.BlockWithValidatorList, len(safeBlocks))
+	for ii, safeBlock := range safeBlocks {
+		safeBlockHash, err := safeBlock.Hash()
+		if err != nil {
+			return nil, errors.Errorf("HandleFastHostStuffBlockProposal: Error hashing safe block: %v", err)
+		}
+
+		utxoViewAtSafeBlock, err := cc.blockchain.getUtxoViewAtBlockHash(*safeBlockHash)
```

Comment:
I think we may be able to avoid constructing a new utxo view for each safeBlockHash as this is probably a relatively expensive operation. I think we can compute the epoch given the block height, which then gives us the snapshot at epoch and then a vanilla UtxoView w/o any connected blocks should be able to fetch the validator set for us. The advantage of this is that we only need a single utxo view and we'll have the validator set cached for each safe block. We shouldn't have more than 2-3 different validator sets and we'll basically be fetching the same information for each block (just depends on the epoch its in really).

Reply:
This is a great optimization! I've added an in-memory cache for this.